### ISSUE TITLE
The range operator needs to be changed

### ISSUE THREAD

**Issue by Silentdoer** [crystal-lang/crystal#13675]
## Bug Report

should use  ..  and  ..= 

Make sure to review these points before submitting issues - thank you!

- Make sure a similar issue does not exist yet: use the search box, search engine and look at [Stack Overflow](https://stackoverflow.com/questions/tagged/crystal-lang).
- **Include reproducible code**: we should be able to paste it into an editor, compile and run it and get the same error as you. Otherwise it's impossible for us to reproduce the bug.
- Don't **only** use `play.crystal-lang.org` or `carc.in`: code might be lost and the issue will remain incomplete. Write code in the issue itself.
- Reduce code, if possible, to the minimum size that reproduces the bug.
- If all of the above is impossible due to a large project, create a branch that reproduces the bug and point us to it.
- Include Crystal compiler version (`crystal -v`) and OS. If possible, try to see if the bug still reproduces on master.

**Comment by Blacksmoke16** [comment]
What's the use case/reasoning?

**Comment by beta-ziliani** [comment]
@Silentdoer I took the liberty of editing your issue to make it more descriptive, hope you don't mind!

**Comment by straight-shoota** [comment]
> Crystal is the only language I make this mistake in.

Ruby is the same (that's where Crystal inherited it from).

**Comment by straight-shoota** [comment]
It could be an option to consider introducing `..<` and `..=` as alternative syntax (and eventual replacement). It would not break anything and be compatible with `..` and `...`, so should be possible.

**Comment by zw963** [comment]
Yes, the ruby usage of `..` and `...` not good, though, I realized this after contacting other languages, because it is easy to confuse, Ruby is my first deep learn programming language, i thought many language were use like this.

I propose add new operator ..= ..<, as described in https://github.com/crystal-lang/crystal/issues/13675#issuecomment-1640701917, it make sense, but never change/drop old behavior.

**Comment by HertzDevil** [comment]
One nice "feature" of the three-dot operator is that code like `foo = ...` or `foo.bar(1, ..., 2)` is perfectly valid syntax, so it won't throw up syntax highlighting in documentation where the dots don't even express a `Range` at all.

**Comment by zw963** [comment]
I never misunderstood `..` with `...`, and when code review, i will notice it specially, because i come from Ruby land.

Yes, we just inherited from ruby, there is no more thinking when design Crystal, but, that not so bad anyway, is Ruby is discussing replace it to use new syntax too? 

Maybe, we need add some document,for users come other land, how to remember, sure, for me, it not necessary, because i familiar with it, but for the new guys, you can try remember like this:

for `..`, just a NORMAL range, that it, includes the end.
for `...`, the `third dot` occupied the position of the end, so, end is not included.

Clear enough?

**Comment by zw963** [comment]
> (on top of the fact that `..` is the opposite of other languages - they _do not_ include the end, furthering the confusion when coming back to Crystal)

Yes, this is the point, the opposite of other languages, so we should consider add new syntax which compatible to other languages, but, leave old behavior works forever.

**Comment by straight-shoota** [comment]
> (on top of the fact that `..` is the opposite of other languages - they *do not* include the end, furthering the confusion when coming back to Crystal)

I'm surprised by this statement. In which languages does `..` denote an *open end* range?

**Comment by oprypin** [comment]
Nim has the syntax `0 .. 9` or `0 ..< 10` (actually originally it was technically `0 .. <10`, but not anymore).
Adding this as an alias to `...` would be fine by me.
`..=` is unnecessary, actually it's worse because it's an asymmetric operator denoting a symmetric operation.
And my topmost preference for how to resolve this issue is: no action. Sometimes "no churn" is better than "slight improvement and huge churn".

**Comment by oprypin** [comment]
@z64 That just seems like.. you could ban all usages of `..` in your team and that's the problem solved.

**Comment by straight-shoota** [comment]
@z64 Python and Odin are not actually a `..` operator, though (arguably, it's just rotated by 90°). They are specifically targeted for subslicing, so I'd think they're similar to the examples with JS and PHP slice methods. Apparently these languages prefer (or expect) open ended ranges for subslicing. The same applies to Zig actually, where `..` is exclusively used for subslicing.
I'm not sure if this is actually reasonable. Looking at the Crystal repo, subslicing operations (`#[Range]` method) much more often use closed end ranges than open ended ones. I haven't looked too deeply into the individual examples, but as a general take away, I'm glad to have the option of choosing either semantic depending on the specific use case.

I found this comparison across multiple languages: http://rigaux.org/language-study/syntax-across-languages/VrsDatTps.html#VrsDatTpsRng
In all languages listed there, `..` designates an inclusive range. Kotlin is another example of that category.

And I really think that makes the most sense for `..`, *especially* when there's operators for both open and closed ranges.

I think Rust and Zig made a very bad choice with `..` meaning open end and `..=`/`...` meaning closed end. They got it the wrong way around! This contradicts the intuitive wisdom of https://github.com/crystal-lang/crystal/issues/13675#issuecomment-1644016280 and deviates from the semantics for `..` which apparently most other languages before them had established.
So if you're confused about whether `..` means inclusive or exclusive, you might just thank Rust and Zig for spoiling the simplicity 😛

I don't know what the reasons are for their syntax decisions. Maybe they have good reasons. But I can't really see any right now. IMO the opposite mapping would've been more intutive and congruent with previous works.

Interestingly, Zig has these two range syntaxes but they are used in completely different contexts: `..` is for slices, `...` for `select` branches. So they're not even directly related or interchangeable. Perhaps they just evolved independently somehow.

Anyway, I'm not saying that the semantics Crystal inherited from Ruby (`..` meaning closed end and `...` meaning open end) are the best solution. But as far as I can tell, it's a much better choice than the inverted semantics in Rust and Zig. Other solutions might be even better, though.
